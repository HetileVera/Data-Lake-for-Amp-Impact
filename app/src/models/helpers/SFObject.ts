import filter from "lodash/filter";
import has from "lodash/has";
import sortBy from "lodash/sortBy";

import { types, Instance } from "mobx-state-tree";
import { values } from "mobx";

import { IRawSalesforceObject } from "api/list-salesforce-objects";
import { defaultObjectsMap } from "data/default-npsp-objects";
import { Field, IField } from "models/helpers/Field";
import { isCompoundType } from "models/helpers/utils";

/**
 * Represents a Salesforce object
 */

const sfProps = {
  name: "",
  label: "",
  fields: types.map(Field),
  fieldsLoaded: false,
  selected: false,
  stale: false,
};

const sfActions = (self: any) => ({
  setObject(rawObject: IRawSalesforceObject) {
    self.fields.clear();
    self.name = rawObject.name;
    self.label = rawObject.label;
    self.fieldsLoaded = false;
    self.selected = false;
    self.stale = false;
  },

  setFields(rawFields: { name: string; label: string; type: string }[]) {
    self.fields.clear();
    for (const rawField of rawFields) {
      // Compound fields are excluded
      self.fields.set(rawField.name, { ...rawField, excluded: isCompoundType(rawField.type) });
    }
  },

  setField(rawField: { name: string; label: string; type: string; excluded: boolean }) {
    self.fields.set(rawField.name, { ...rawField, excluded: rawField.excluded || isCompoundType(rawField.type) });
  },

  setFieldsLoaded(flag: boolean) {
    self.fieldsLoaded = flag;
  },

  setSelected(flag: boolean) {
    self.selected = flag;
  },

  markStale() {
    self.stale = true;
  },
});

const sfViews = (self: any) => ({
  get selectableFields(): IField[] {
    // Returns the list of fields that can be selected
    const filtered = filter(values(self.fields), ["canExclude", true]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get listableFields(): IField[] {
    // listable fields are all the fields with the exception of the compound field
    const filtered = filter(values(self.fields), ["compound", false]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get immutableFields(): IField[] {
    const filtered = filter(values(self.fields), { compound: false, canExclude: false });
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get selectedFields(): IField[] {
    // Returns all fields that are included (including immutable fields)
    const filtered = filter(values(self.fields), ["excluded", false]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get userSelectedFields(): IField[] {
    // Returns all fields that are included (excluded immutable fields)
    const filtered = filter(values(self.fields), { excluded: false, canExclude: true });
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get excludedFields(): IField[] {
    // Returns excluded fields (with the exception of compound fields)
    const filtered = filter(values(self.fields), { excluded: true, compound: false });
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get allExcludedFields(): IField[] {
    // Returns all excluded fields including the compound field
    const filtered = filter(values(self.fields), ["excluded", true]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },
  get fieldsCount(): number {
    // All fields (selected or not) excluding compound fields
    return self.listableFields.length;
  },

  get selectedFieldsCount(): number {
    const result = self.selectedFields;
    return result.length;
  },

  get excludedFieldsCount(): number {
    const result = self.excludedFields;
    return result.length;
  },

  get excluded() {
    // Returns the excluded fields as expected by the step functions lambdas. The following is an example of the
    // expected exclude:
    // "exclude": {
    //   "CompoundField__c": { "type": "location", "label": "<field label>", "$comment": "Compound Field" },
    //   "ABC__c": { "type": "string", "label", "<field label>", "$comment": "ABC field" },
    //   "...": { ... },
    // },
    const excludedFields = self.allExcludedFields;
    const exclude: { [index: string]: { type: string; label: string } } = {};
    for (const field of excludedFields) {
      exclude[field.name] = {
        type: field.type,
        label: field.label,
      };
    }

    return exclude;
  },

  get isDefault(): boolean {
    // Is this object one of the default NPSP objects?
    return has(defaultObjectsMap, self.name);
  },
});

export const SFObject = types.model("SFObject", sfProps).actions(sfActions).views(sfViews);

export interface ISFObject extends Instance<typeof SFObject> {}
