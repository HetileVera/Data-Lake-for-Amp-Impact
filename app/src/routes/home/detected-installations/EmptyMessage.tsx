import React from "react";
import { Box, Button } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useDetectedInstallationStore } from "models/useDetectedInstallationStore";
import { useColorScheme } from "models/useColorScheme";

interface IEmptyMessage {
  onCancel: () => void;
}

export const EmptyMessage = observer(({ onCancel }: IEmptyMessage) => {
  const { tone, colorScheme } = useColorScheme();
  const { store, isError, isLoading, isReloading, isReady } = useDetectedInstallationStore();
  if (isError || isLoading || isReloading || !isReady) return null;
  if (!store.empty) return null;

  return (
    <>
      <Box bg={tone(100)} textAlign="center" p={4} fontSize="sm" color={tone(600)} borderRadius="md">
        No data lakes detected in AWS account # {store.credentials.accountId}
      </Box>
      <Box textAlign="right" w="full" mt={6} mb={0}>
        <Button colorScheme={colorScheme} size="sm" variant="outline" onClick={onCancel}>
          Ok
        </Button>
      </Box>
    </>
  );
});
