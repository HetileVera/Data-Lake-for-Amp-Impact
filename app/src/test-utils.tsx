import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";

import { theme } from "themes/orange";

export const CustomChakraProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={theme}>
      <div data-testid="test-root">{children}</div>
    </ChakraProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) => render(ui, { wrapper: CustomChakraProvider, ...options });

export * from "@testing-library/react";

export { customRender as render };
