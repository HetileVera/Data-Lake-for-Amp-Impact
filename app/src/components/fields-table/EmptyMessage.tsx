import React from "react";
import isEmpty from "lodash/isEmpty";
import { Box } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useColorScheme } from "models/useColorScheme";
import { FieldsViewOptions } from "models/useFieldsTable";
import { ISFObject } from "models/helpers/SFObject";
import { IField } from "models/helpers/Field";

interface IEmptyMessage {
  object: ISFObject;
  viewOption: FieldsViewOptions;
  fields: IField[];
  searchText: string;
}

export const EmptyMessage = observer(({ object, viewOption, fields, searchText }: IEmptyMessage) => {
  const { tone } = useColorScheme();

  if (fields.length > 0) return null;

  let content = "No fields are found";
  let size = 0;

  if (!isEmpty(searchText)) content = "No matches are found";
  else {
    switch (viewOption) {
      case FieldsViewOptions.All:
        size = object.immutableFields.length;
        if (size > 0) {
          content = `There are ${size} included fields. They are mandatory.`;
        } else {
          content = "No fields are found";
        }
        break;
      case FieldsViewOptions.Only_Included:
        size = object.immutableFields.length;
        if (size > 0) {
          content = `There are ${size} included fields. They are mandatory.`;
        } else {
          content = "No included fields are found";
        }
        break;
      case FieldsViewOptions.Only_Excluded:
        content = "No excluded fields are found";
    }
  }

  return (
    <Box bg={tone(100)} textAlign="center" p={4} fontSize="xs" color={tone(600)} borderRadius="md">
      {content}
    </Box>
  );
});
