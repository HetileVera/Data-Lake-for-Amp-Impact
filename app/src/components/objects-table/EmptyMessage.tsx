import React from "react";
import isEmpty from "lodash/isEmpty";
import { Box } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useColorScheme } from "models/useColorScheme";
import { ObjectsViewOptions } from "models/useObjectsTable";
import { ISFObject } from "models/helpers/SFObject";

interface IEmptyMessage {
  viewOption: ObjectsViewOptions;
  objects: ISFObject[];
  searchText: string;
}

export const EmptyMessage = observer(({ viewOption, objects, searchText }: IEmptyMessage) => {
  const { tone } = useColorScheme();

  if (objects.length > 0) return null;
  let content = "No objects are found";

  if (!isEmpty(searchText)) content = "No matches are found";
  else {
    switch (viewOption) {
      case ObjectsViewOptions.All:
        content = "No objects are found";
        break;
      case ObjectsViewOptions.Only_Included:
        content = "No included objects are found";
        break;
      case ObjectsViewOptions.Only_Excluded:
        content = "No excluded objects are found";
    }
  }

  return (
    <Box bg={tone(100)} textAlign="center" p={4} fontSize="xs" color={tone(600)} borderRadius="md">
      {content}
    </Box>
  );
});
