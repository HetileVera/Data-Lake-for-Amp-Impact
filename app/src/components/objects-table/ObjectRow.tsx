import React from "react";
import isEmpty from "lodash/isEmpty";
import { Badge, Box, Button, Flex, Icon, Tr, Td, keyframes, Alert, AlertTitle, AlertDescription, CloseButton } from "@chakra-ui/react";
import { BsCircle, BsCheckCircleFill } from "react-icons/bs";
import { observer } from "mobx-react";
import { SpinnerIcon } from "@chakra-ui/icons";

import { niceNumber } from "helpers/utils";
import { useMetadataStore } from "models/MetadataStore";
import { useColorScheme } from "models/useColorScheme";
import { FieldsTable } from "components/fields-table/FieldsTable";
import { ISFObject } from "models/helpers/SFObject";

interface IObjectRow {
  object: ISFObject;
  isProcessing: boolean;
  setProcessing: React.Dispatch<React.SetStateAction<boolean>>;
}

export const ObjectRow = observer(({ object, isProcessing, setProcessing }: IObjectRow) => {
  const { tone, colorScheme } = useColorScheme();
  const { store } = useMetadataStore();
  const [isLoadingFields, setLoadingField] = React.useState(false);
  const [showFields, setShowFields] = React.useState(false);
  const [error, setError] = React.useState("");
  const hasError = !isEmpty(error);

  const onToggleFields = async () => {
    setShowFields((previous) => !previous);
  };

  const onAddOrRemove = async () => {
    if (isProcessing || showFields) return;

    setError("");
    const isSelected = object.selected;

    if (isSelected) {
      object.setSelected(false);
      return;
    }

    if (object.fieldsLoaded) {
      object.setSelected(true);
      return;
    }

    setProcessing(true);
    setLoadingField(true);

    try {
      await store.selectObject(object);
    } catch (err: any) {
      console.error(err);
      setError(err.message);
    } finally {
      setProcessing(false);
      setLoadingField(false);
    }
  };

  let props: any = {
    color: isProcessing && !isLoadingFields ? tone(100) : tone(500),
    fontSize: "large",
    mt: 1,
    ml: 3,
    mr: 0,
  };

  const frames = keyframes`to { transform: rotate(360deg); }`;

  if (isLoadingFields) {
    props = { ...props, color: tone(500), as: SpinnerIcon, animation: `${frames} 1s linear infinite` };
  } else if (object.selected) {
    props = { ...props, as: BsCheckCircleFill };
  } else {
    props = { ...props, as: BsCircle };
  }

  const fontColor = isProcessing && !isLoadingFields ? tone(100) : tone(500);

  return (
    <>
      <Tr cursor={isProcessing || showFields ? "default" : "pointer"}>
        <Td colSpan={1}>
          <Flex>
            <Icon {...props} onClick={onAddOrRemove} />
            <Box flex={1} ml={4} color={fontColor} onClick={onAddOrRemove}>
              <Box fontWeight={object.selected ? "semibold" : "normal"}>
                {object.label}
                {object.isDefault && (
                  <Badge
                    colorScheme={colorScheme}
                    borderRadius="md"
                    bg={isProcessing && !isLoadingFields ? tone(50) : tone(75)}
                    color={fontColor}
                    p={0}
                    pl={1}
                    pr={1}
                    mt={0}
                    ml={2}
                    fontSize="0.50rem"
                    fontWeight="normal"
                  >
                    npsp
                  </Badge>
                )}
              </Box>
              <Box fontSize="0.6rem">{object.name}</Box>
            </Box>
          </Flex>
        </Td>
        <Td isNumeric>{object.selected && <>{niceNumber(object.selectedFieldsCount)}</>}</Td>
        <Td textAlign="right">
          {object.selected && !hasError && (
            <Button size="xs" fontSize="0.6rem" variant="outline" colorScheme={colorScheme} isDisabled={isProcessing} onClick={onToggleFields} mt={1}>
              {!showFields && <>Customize Fields</>}
              {showFields && <>Collapse Fields Table</>}
            </Button>
          )}
        </Td>
      </Tr>
      {hasError && (
        <Tr>
          <Td colSpan={3}>
            <Alert status="error" variant="left-accent" mt={0} mb={10} color="red.700" alignItems="flex-start" fontSize="sm">
              <Box flex="1">
                <AlertTitle>Error</AlertTitle>
                <AlertDescription display="block">{error}</AlertDescription>
              </Box>
              <CloseButton position="absolute" right="8px" top="8px" onClick={() => setError("")} />
            </Alert>
          </Td>
        </Tr>
      )}
      {showFields && (
        <Tr>
          <Td colSpan={3} pb={12} pt={6}>
            <FieldsTable object={object} />
          </Td>
        </Tr>
      )}
    </>
  );
});
