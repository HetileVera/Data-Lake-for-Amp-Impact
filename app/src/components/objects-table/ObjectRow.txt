import { cleanup, screen } from "@testing-library/react";
import { create } from "react-test-renderer";

import { render, RenderWithChakra } from "test-utils";

import { SFObject } from "models/helpers/SFObject";

import { ObjectRow } from "components/objects-table/ObjectRow";
import { AppStore } from "models/AppStore";
import { initializeStore } from "AppContext";

// import * as metadataStoreHooks from "models/MetadataStore";

describe("ObjectRow component", () => {
  beforeEach(() => {
    const appStore = AppStore.create({});
    const installation = appStore.newInstallation();
    appStore.setInstallation(installation);

    initializeStore();

    // jest.spyOn(metadataStoreHooks, "useMetadataStore").mockImplementation(() => ({
    //   isError: false,
    //   isReady: false,
    //   store: installation.metadataStore,
    // }));
    // jest.spyOn(installationHooks, "useInstallation").mockImplementation(() => {
    //   return installation;
    // });
  });

  afterEach(cleanup);

  test("snapshot", () => {
    const sfModel = SFObject.create({
      name: "TestField",
      label: "Test Field",
      fields: {},
      fieldsLoaded: false,
      selected: false,
      stale: false,
    });

    const tree = create(
      <RenderWithChakra>
        <ObjectRow object={sfModel} isProcessing={true} setProcessing={() => jest.fn()} />
      </RenderWithChakra>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
