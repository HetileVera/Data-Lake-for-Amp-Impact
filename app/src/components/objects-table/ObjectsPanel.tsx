import React from "react";
import isEmpty from "lodash/isEmpty";
import { Box, Flex, IconButton, Input, InputGroup, InputRightElement, Table, Tbody, Thead, Th, Tr } from "@chakra-ui/react";
import { observer } from "mobx-react";
import { SmallCloseIcon } from "@chakra-ui/icons";

import { useMetadataStore } from "models/MetadataStore";
import { useColorScheme } from "models/useColorScheme";
import { useObjectsTable } from "models/useObjectsTable";
import { PaginationButtons } from "components/PaginationButtons";

import { ObjectsViewOptionsPanel } from "./ObjectsViewOptionsPanel";
import { EmptyMessage } from "./EmptyMessage";
import { ObjectRow } from "./ObjectRow";

export const ObjectsPanel = observer(() => {
  const { tone, colorScheme } = useColorScheme();
  const { store } = useMetadataStore();
  const [isProcessing, setProcessing] = React.useState(false);
  const { objects, viewOption, setViewOption, totalPages, currentPage, setCurrentPage, searchText, setSearchText, totalMatches } = useObjectsTable({
    store,
    pageSize: 25,
  });
  const handleTextChange: React.ChangeEventHandler<HTMLInputElement> = (event) => setSearchText(event.target.value);

  return (
    <Box mt={6} mb={4}>
      <ObjectsViewOptionsPanel viewOption={viewOption} setViewOption={setViewOption} isProcessing={isProcessing} />
      {(totalPages > 1 || !isEmpty(searchText)) && (
        <Box position="relative" mb={8}>
          <Flex mb={0}>
            <InputGroup size="sm">
              <Input type="text" placeholder="Search" borderRadius="md" value={searchText} onChange={handleTextChange} isDisabled={isProcessing} />
              <InputRightElement>
                <IconButton
                  onClick={() => setSearchText("")}
                  size="sm"
                  variant="ghost"
                  color="purple.200"
                  colorScheme={colorScheme}
                  aria-label="Clear search text"
                  _active={{ bg: "none", color: tone(900) }}
                  _hover={{ bg: "none", color: tone(500) }}
                  _focus={{ boxShadow: "none" }}
                  icon={<SmallCloseIcon fontSize="lg" />}
                  isDisabled={isProcessing}
                />
              </InputRightElement>
            </InputGroup>
            <PaginationButtons totalPages={totalPages} currentPage={currentPage} setCurrentPage={setCurrentPage} isProcessing={isProcessing} />
          </Flex>
          <Box p={0} m={0} fontSize="0.6rem" position="absolute" bottom="-20px" left="10px">
            {!isEmpty(searchText) && (
              <>
                Found <b>{totalMatches}</b> matches
              </>
            )}
          </Box>
        </Box>
      )}

      <EmptyMessage objects={objects} viewOption={viewOption} searchText={searchText} />

      {objects.length > 0 && (
        <>
          <Table variant="simple" colorScheme={colorScheme} size="sm" mb={6}>
            <Thead>
              <Tr>
                <Th pl={10}>Objects</Th>
                <Th w="100px" textAlign="right">
                  Fields
                </Th>
                <Th w="170px" isNumeric />
              </Tr>
            </Thead>
            <Tbody>
              {objects.map((object) => (
                <ObjectRow key={`${object.label}-${object.name}`} object={object} isProcessing={isProcessing} setProcessing={setProcessing} />
              ))}
            </Tbody>
          </Table>
          <Flex mb={0} direction="row-reverse">
            <PaginationButtons totalPages={totalPages} currentPage={currentPage} setCurrentPage={setCurrentPage} isProcessing={isProcessing} />
          </Flex>
        </>
      )}
    </Box>
  );
});
