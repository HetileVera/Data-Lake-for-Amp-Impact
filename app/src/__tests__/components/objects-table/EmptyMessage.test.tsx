import { cleanup, screen } from "@testing-library/react";
import { create } from "react-test-renderer";

import { render, CustomChakraProvider } from "test-utils";

import { SFObject } from "models/helpers/SFObject";

import { EmptyMessage } from "components/objects-table/EmptyMessage";
import { ObjectsViewOptions } from "models/useObjectsTable";

describe("ObjectsTable -> EmptyMessage component", () => {
  afterEach(cleanup);

  test("snapshot", () => {
    const sfModel = SFObject.create({
      name: "TestField",
      label: "Test Field",
      fields: {},
      fieldsLoaded: false,
      selected: false,
      stale: false,
    });

    const tree = create(
      <CustomChakraProvider>
        <EmptyMessage objects={[sfModel]} viewOption={ObjectsViewOptions.All} searchText="" />
      </CustomChakraProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("render null if objects prop is not empty", () => {
    const testDiv = document.createElement("div");
    testDiv.id = "root";

    const sfModel = SFObject.create({
      name: "TestField",
      label: "Test Field",
      fields: {},
      fieldsLoaded: false,
      selected: false,
      stale: false,
    });

    render(<EmptyMessage objects={[sfModel]} viewOption={ObjectsViewOptions.All} searchText="" />, {
      container: document.body.appendChild(testDiv),
    });

    // eslint-disable-next-line testing-library/no-node-access
    expect(document.getElementById("root")).toBeInTheDocument();
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.getElementById("root")).toHaveTextContent("");
  });

  test("render no mathces are found if search text is not empty", () => {
    render(<EmptyMessage objects={[]} viewOption={ObjectsViewOptions.All} searchText="test-search-text" />);

    expect(screen.getByText(/no matches are found/i)).toBeInTheDocument();
  });

  test("render no objects are found", () => {
    render(<EmptyMessage objects={[]} viewOption={ObjectsViewOptions.All} searchText="" />);

    expect(screen.getByText(/no objects are found/i)).toBeInTheDocument();
  });
});
