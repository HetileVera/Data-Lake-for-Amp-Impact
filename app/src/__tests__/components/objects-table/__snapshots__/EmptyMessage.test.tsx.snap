// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectsTable -> EmptyMessage component snapshot 1`] = `
Array [
  <div
    data-testid="test-root"
  />,
  <span />,
]
`;
